yarn-project/ethereum/src/gas_utils.ts

```diff
@@ -0,0 +1,216 @@
+    return retry(
+      async () => {
+        const gas = await estimator();
+        return gas + (gas * this.gasConfig.bufferPercentage) / 100n;
+      },
+      'gas estimation',
+      makeBackoff([1, 2, 3]),
+      this.logger,
+      false,
+    );
+  }
+
+  /**
+   * Monitors a transaction and attempts to speed it up if it gets stuck
+   * @param txHash - The hash of the transaction to monitor
+   * @param walletClient - The wallet client for sending replacement tx
+   * @param originalTx - The original transaction data for replacement
+   * @param config - Monitor configuration
+   */
+  public async monitorTransaction(
+    txHash: Hex,
+    walletClient: WalletClient,
+    originalTx: {
+      to: Address;
+      data: Hex;
+      nonce: number;
+      gasLimit: bigint;
+      maxFeePerGas: bigint;
+    },
+    txMonitorConfig?: TransactionMonitorConfig,
+  ): Promise<TransactionReceipt> {
+    const config = { ...this.monitorConfig, ...txMonitorConfig };
+    let attempts = 0;
+    let lastSeen = Date.now();
+    let currentTxHash = txHash;
+
+    // Flag to use for first run to avoid logging errors.
+    let firstRun = true;
+
+    while (true) {
+      try {
+        // Check transaction status
+        const receipt = await this.publicClient.getTransactionReceipt({ hash: currentTxHash });
+        if (receipt) {
+          this.logger?.info(`Transaction ${currentTxHash} confirmed`);
+          return receipt;
+        }
+
+        if (firstRun) {
+          firstRun = false;
+          // Wait a second for tx to be broadcast
+          await sleep(1000);
+        }
+
+        // Check if transaction is pending
+        const tx = await this.publicClient.getTransaction({ hash: currentTxHash });
+        this.logger?.info(`Transaction ${currentTxHash} pending`);
+        if (tx) {
+          lastSeen = Date.now();
+          await new Promise(resolve => setTimeout(resolve, config.checkIntervalMs));
+          continue;
+        }
+
+        // Transaction not found and enough time has passed - might be stuck
+        if (Date.now() - lastSeen > config.stallTimeMs && attempts < config.maxAttempts) {
+          attempts++;
+          const newGasPrice = (originalTx.maxFeePerGas * (100n + config.gasPriceIncrease)) / 100n;
+
+          this.logger?.info(
+            `Transaction ${currentTxHash} appears stuck. Attempting speed-up ${attempts}/${config.maxAttempts} ` +
+              `with new gas price ${formatGwei(newGasPrice)} gwei`,
+          );
+
+          // Send replacement transaction with higher gas price
+          const account = await walletClient.getAddresses().then(addresses => addresses[0]);
+          currentTxHash = await walletClient.sendTransaction({
+            chain: null,
+            account,
+            to: originalTx.to,
+            data: originalTx.data,
+            nonce: originalTx.nonce,
+            gas: originalTx.gasLimit,
+            maxFeePerGas: newGasPrice,
```

Comment:
I'm not sure if bumping only `maxFeePerGas` without touching `maxPriorityFeePerGas` has the intended effect. If `maxPriorityFeePerGas` is not set, then viem defaults it to either a chain default or to whatever the json rpc interface returns (see [here](https://github.com/wevm/viem/blob/main/src/actions/public/estimateMaxPriorityFeePerGas.ts)). So we're not actually bumping the priority fee, we're just setting it to whatever nodes say is the latest value to use, and ignoring the `gasPriceIncrease`.

Maybe we should be setting `maxPriorityFeePerGas` as well? I'm thinking we could just always set `maxFeePerGas` to `gasConfig.maxGwei`, and then just handle stuck txs by bumping the `maxPriorityFeePerGas`, since the user never gets charged more than `baseFee + priorityFee`. We could get the max prio fee from either `eth_maxPriorityFeePerGas` RPC as viem does, or from a gas price oracle like [BlockNative](https://www.blocknative.com/gas-platform). Then, for every successive retry, we bump the maxPrioFee to the new `eth_maxPriorityFeePerGas` we see, or to a multiplier of the previous prio fee, whatever's greater.